eventuate-messaging-kafka-basic-consumer/src/test/java/io/eventuate/messaging/kafka/basic/consumer/EventuateKafkaConsumerTest.java

```diff
@@ -0,0 +1,229 @@
+    when(mockedHandler.apply(any(), any())).thenThrow(new RuntimeException("Something happened!"));
+    createConsumer(mockedHandler);
+    sendMessage();
+    assertRecordHandled();
+    assertMessageReceivedByNewConsumer();
+  }
+
+  @Test
+  public void testConsumerSwitchOnHanging() throws InterruptedException {
+    createConsumer(mockedHandler);
+    sendMessage();
+    assertRecordHandled();
+    assertMessageReceivedByNewConsumer();
+  }
+
+  @Test
+  public void testConsumerStop() throws InterruptedException {
+    EventuateKafkaConsumer eventuateKafkaConsumer = createConsumer(mockedHandler);
+    sendMessage();
+    assertRecordHandled();
+    eventuateKafkaConsumer.stop();
+    assertMessageReceivedByNewConsumer();
+  }
+
+  @Test
+  public void testNotClosedConsumerOnStop() throws InterruptedException {
+    eventuateKafkaConsumerConfigurationProperties.getProperties().put("max.poll.interval.ms", "1000");
+    EventuateKafkaConsumer eventuateKafkaConsumer = createConsumer(mockedHandler);
+    sendMessage();
+    assertRecordHandled();
+    eventuateKafkaConsumer.closeConsumerOnStop = false;
+    eventuateKafkaConsumer.stop();
+    assertMessageReceivedByNewConsumer();
+  }
+
+  @Test
+  public void testKafkaTimeOut() throws Exception {
+    eventuateKafkaConsumerConfigurationProperties.getProperties().put("default.api.timeout.ms", "3000");
+
+    EventuateKafkaConsumer eventuateKafkaConsumer = createNoOpConsumer();
+
+    Runnable onCommitOffsetsFailedCallback = Mockito.mock(Runnable.class);
+
+    ConsumerCallbacks consumerCallbacks = new ConsumerCallbacks() {
+      @Override
+      public void onTryCommitCallback() {
+        if (!isItCalledFromPollingLoop()) {
+          return;
+        }
+
+        try {
+          stopKafka();
+          Thread.sleep(3000);
+        } catch (Exception e) {
+          throw new RuntimeException(e);
+        }
+      }
+
+      @Override
+      public void onCommitedCallback() {
+        if (!isItCalledFromPollingLoop()) {
+          return;
+        }
+
+        eventuateKafkaConsumer.stop();
+        sendMessage("test-value-2");
+      }
+
+      @Override
+      public void onCommitFailedCallback() {
+        onCommitOffsetsFailedCallback.run();
+      }
+
+      private boolean isItCalledFromPollingLoop() {
+        return Arrays
+                .stream(Thread.currentThread().getStackTrace())
+                .anyMatch(stackTraceElement -> stackTraceElement.getMethodName().equals("runPollingLoop"));
+      }
+    };
+
+    eventuateKafkaConsumer.consumerCallbacks = Optional.of(consumerCallbacks);
+
+    sendMessage();
+
+    Eventually.eventually(60, 500, TimeUnit.MILLISECONDS, () -> {
+      Mockito.verify(onCommitOffsetsFailedCallback).run();
+    });
+
+    startKafka();
+
+    assertMessageReceivedByNewConsumer("test-value-2");
+  }
+
+  private void stopKafka() throws IOException {
+    Runtime.getRuntime().exec("docker-compose stop kafka");
```

Comment:
Strictly speaking I don't think you can ignore the Process object returned by exec().
See https://docs.oracle.com/javase/7/docs/api/java/lang/Process.html